funcx_endpoint/funcx_endpoint/endpoint/endpoint_manager.py

```diff
@@ -0,0 +1,508 @@
+                "stderr": config.stderr,
+                "environment": config.environment,
+                "funcx_service_address": config.funcx_service_address,
+            },
+        }
+
+    def request_shutdown(self, sig_num, curr_stack_frame):
+        self._time_to_stop = True
+
+    def set_child_died(self, sig_num, curr_stack_fframe):
+        self._wait_for_child = True
+
+    def wait_for_children(self):
+        try:
+            self._wait_for_child = False
+            wflags = os.WEXITED | os.WCONTINUED | os.WSTOPPED | os.WNOHANG
+            siginfo = os.waitid(os.P_ALL, 0, wflags)
+            while siginfo is not None:
+                pid = siginfo.si_pid
+                if os.WIFEXITED(siginfo.si_status) or siginfo.si_code == os.CLD_EXITED:
+                    *_, proc_args = self._child_args.pop(pid, (None, None, None, None))
+                    rc = os.WEXITSTATUS(siginfo.si_status)
+                    if rc == 0 and proc_args:
+                        log.info(
+                            f"User endpoint stopped normally ({pid}) [{proc_args}]"
+                        )
+                    elif rc:
+                        if proc_args:
+                            log.warning(
+                                f"User endpoint return code: {rc} ({pid}) [{proc_args}]"
+                            )
+                        else:
+                            log.warning(f"Command return code: {rc} ({pid})")
+                elif os.WIFSIGNALED(siginfo.si_status):
+                    termsig = os.WTERMSIG(siginfo.si_status)
+                    *_, proc_args = self._child_args.pop(pid, (None, None, None, None))
+                    if proc_args:
+                        log.warning(
+                            f"User endpoint terminated by signal:"
+                            f" {termsig} ({pid}) [{proc_args}]"
+                        )
+                    else:
+                        log.warning(f"Command terminated by signal: {termsig} ({pid})")
+                else:
+                    *_, proc_args = self._child_args.get(pid, (None, None, None, None))
+                    proc_args = f" [{proc_args}]" if proc_args else ""
+                    log.warning(
+                        "Unhandled path: indeterminate child process state."
+                        f" ({pid}){proc_args}"
+                    )
+                siginfo = os.waitid(os.P_ALL, 0, wflags)
+
+        except ChildProcessError:
+            pass
+        except Exception as e:
+            log.exception(f"Failed to wait for a child process: {e}")
+
+    def _install_signal_handlers(self):
+        signal.signal(signal.SIGTERM, self.request_shutdown)
+        signal.signal(signal.SIGINT, self.request_shutdown)
+        signal.signal(signal.SIGQUIT, self.request_shutdown)
+
+        signal.signal(signal.SIGCHLD, self.set_child_died)
+
+    def start(self):
+        log.info("\n\n========== Endpoint Manager begins ==========")
```

Comment:
This is a temporary debug statement, right?  Just a reminder for removal later.

Reply:
Had a good convo off-PR.  Agreed with Kevin's overall logging intention, was just one comment about combining the two log statements here into one for more conciseness.  Will resolve the others for now.